python/ray/data/impl/progress_bar.py

```diff
@@ -54,18 +59,29 @@ def __init__(self, name: str, total: int, position: int = 0):
             self._bar = None
 
     def block_until_complete(self, remaining: List[ObjectRef]) -> None:
+        t = threading.current_thread()
         while remaining:
-            done, remaining = ray.wait(remaining, fetch_local=False)
+            done, remaining = ray.wait(remaining, fetch_local=False, timeout=0.1)
             self.update(len(done))
 
+            with _canceled_threads_lock:
+                if t in _canceled_threads:
+                    break
+
     def fetch_until_complete(self, refs: List[ObjectRef]) -> List[Any]:
         ref_to_result = {}
         remaining = refs
+        t = threading.current_thread()
         while remaining:
-            done, remaining = ray.wait(remaining, fetch_local=True)
+            done, remaining = ray.wait(remaining, fetch_local=True, timeout=0.1)
             for ref, result in zip(done, ray.get(done)):
```

Comment:
Oh, won't this trigger a `ray.get()` every 100ms? I know that this should be a shared memory reference after the initial fetch, but is there anything else in the `ray.get()` path that's expensive?

Reply:
I think once every 100ms is okay but we can revisit it if it becomes a bottleneck.